src/probability/martingale/convergence.lean

```diff
@@ -0,0 +1,299 @@
+      { simp_rw snorm_one_eq_lintegral_nnnorm at hbdd,
+        intro n,
+        refine (lintegral_mono _ : ∫⁻ ω, ∥f n ω - a∥₊ ∂μ ≤ ∫⁻ ω, ∥f n ω∥₊ + ∥a∥₊ ∂μ).trans _,
+        { intro ω,
+          simp_rw [sub_eq_add_neg, ← nnnorm_neg a, ← ennreal.coe_add, ennreal.coe_le_coe],
+          exact nnnorm_add_le _ _ },
+        { simp_rw [ lintegral_add_right _ measurable_const, lintegral_const],
+          exact add_le_add (hbdd _) le_rfl } },
+      refine ne_of_lt (supr_lt_iff.2 ⟨R + ∥a∥₊ * μ set.univ, ennreal.add_lt_top.2
+          ⟨ennreal.coe_lt_top, ennreal.mul_lt_top ennreal.coe_lt_top.ne (measure_ne_top _ _)⟩,
+          λ n, le_trans _ (hR' n)⟩),
+      refine lintegral_mono (λ ω, _),
+      rw [ennreal.of_real_le_iff_le_to_real, ennreal.coe_to_real, coe_nnnorm],
+      by_cases hnonneg : 0 ≤ f n ω - a,
+      { rw [lattice_ordered_comm_group.pos_of_nonneg _ hnonneg,
+          real.norm_eq_abs, abs_of_nonneg hnonneg] },
+      { rw lattice_ordered_comm_group.pos_of_nonpos _ (not_le.1 hnonneg).le,
+        exact norm_nonneg _ },
+      { simp only [ne.def, ennreal.coe_ne_top, not_false_iff] } },
+    { simp only [hab, ne.def, ennreal.of_real_eq_zero, sub_nonpos, not_le] } },
+  { simp only [hab, ne.def, ennreal.of_real_eq_zero, sub_nonpos, not_le, true_or]},
+  { simp only [ne.def, ennreal.of_real_ne_top, not_false_iff, true_or] }
+end
+
+lemma submartingale.upcrossings_ae_lt_top
+  (hf : submartingale f ℱ μ) (hbdd : ∀ n, snorm (f n) 1 μ ≤ R) :
+  ∀ᵐ ω ∂μ, ∀ a b : ℚ, a < b → upcrossings a b f ω < ∞ :=
+begin
+  simp only [ae_all_iff, eventually_imp_distrib_left],
+  rintro a b hab,
+  exact hf.upcrossings_ae_lt_top' hbdd (rat.cast_lt.2 hab),
+end
+
+/-- An L¹-bounded submartingale converges almost everywhere. -/
+lemma submartingale.exists_ae_tendsto_of_bdd
+  (hf : submartingale f ℱ μ) (hbdd : ∀ n, snorm (f n) 1 μ ≤ R) :
+  ∀ᵐ ω ∂μ, ∃ c, tendsto (λ n, f n ω) at_top (𝓝 c) :=
+begin
+  filter_upwards [hf.upcrossings_ae_lt_top hbdd, liminf_at_top_ae_bdd_of_snorm_bdd
+    (λ n, (hf.strongly_measurable n).measurable.mono (ℱ.le n) le_rfl) hbdd] with ω h₁ h₂,
+  exact tendsto_of_uncrossing_lt_top h₂ h₁,
+end
+
+lemma submartingale.exists_ae_trim_tendsto_of_bdd
+  (hf : submartingale f ℱ μ) (hbdd : ∀ n, snorm (f n) 1 μ ≤ R) :
+  ∀ᵐ ω ∂(μ.trim (Sup_le (λ m ⟨n, hn⟩, hn ▸ ℱ.le _) : (⨆ n, ℱ n) ≤ m0)),
+    ∃ c, tendsto (λ n, f n ω) at_top (𝓝 c) :=
+begin
+  rw [ae_iff, trim_measurable_set_eq],
+  { exact hf.exists_ae_tendsto_of_bdd hbdd },
+  { exact measurable_set.compl (@measurable_set_exists_tendsto _ _ _ _ _ _ (⨆ n, ℱ n) _ _ _ _ _
+    (λ n, ((hf.strongly_measurable n).measurable.mono (le_Sup ⟨n, rfl⟩) le_rfl))) }
+end
+
+/-- **Almost everywhere martingale convergence theorem**: An L¹-bounded submartingale converges
+almost everywhere to a L¹-function which is measurable with respect to `⨆ n, ℱ n`. -/
+lemma submartingale.exists_mem_ℒ1_ae_tendsto_of_bdd
+  (hf : submartingale f ℱ μ) (hbdd : ∀ n, snorm (f n) 1 μ ≤ R) :
```

Comment:
can you introduce a 
```lean
def submartingale.lim :=
if h : ∃ g : Ω → ℝ, strongly_measurable[⨆ n, ℱ n] g ∧
  ∀ᵐ ω ∂μ, tendsto (λ n, f n ω) at_top (𝓝 (g ω)) then some h else 0
```
and formulate your theorem using it? This will be more convenient for the other theorems to have this function, and say that f n converges to it ae / in L^p depending on the assumptions.

Reply:
I've called it `limit_process` instead since the definition doesn't require `submartingale`